### PR TITLE
Fix Hidive - Subtitles and Age-restriction.

### PR DESCRIPTION
**Pull request by chexxor**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Currently the Hidive extractor is broken:
- It doesn't extract subtitles.
- It doesn't select a Hidive profile, if the user has multiple profiles defined. This applies age-restriction to the content.

After login, Hidive redirects your browser to a profile-selection page. On that page, you need to click the button/icon you want to use, which sends a JSON request to the server which sets the profile id as a cookie.

Also, the subtitle information is no longer available (from what I've seen) in a "data-captions" element on the page, but they _are_available in the "rendition" data structure that Hidive gives us.

Fixes:
- https://github.com/yt-dlp/yt-dlp/issues/4279
- https://github.com/yt-dlp/yt-dlp/issues/408

### Before submitting a *pull request* make sure you have:
- [ ] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [ ] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/hidive.py [hunk 1]
```diff
@@ -60,24 +76,22 @@ def _call_api(self, video_id, title, key, data={}, **kwargs):
...
     def _get_subtitles(self, url, video_id, title, key, parsed_urls):
-        webpage = self._download_webpage(url, video_id, fatal=False) or ''
         subtitles = {}
-        for caption in set(re.findall(r'data-captions=\"([^\"]+)\"', webpage)):
-            renditions = self._call_api(
-                video_id, title, key, {'Captions': caption}, fatal=False,
-                note=f'Downloading {caption} subtitle information').get('renditions') or {}
-            for rendition_id, rendition in renditions.items():
-                self._extract_subtitles_from_rendition(rendition, subtitles, parsed_urls)
+        webpage = self._download_webpage(url, video_id, fatal=False) or ''
```
Comment (pukkandan) [inline yt_dlp/extractor/hidive.py:81]: why? unused

File: yt_dlp/extractor/hidive.py [hunk 2]
```diff
@@ -60,24 +76,22 @@ def _call_api(self, video_id, title, key, data={}, **kwargs):
...
+
+        settings = self._call_api(video_id, title, key)
+        for rendition_id, rendition in settings['renditions'].items():
+            audio, version, extra = rendition_id.split('_')
+
+            for cc_file in rendition.get('ccFiles', []):
+                cc_url = url_or_none(try_get(cc_file, lambda x: x[2]))
+                # name is used since we cant distinguish subs with same language code
```
Comment (pukkandan) [inline yt_dlp/extractor/hidive.py:89]: pls explain

File: yt_dlp/extractor/hidive.py [hunk 3]
```diff
@@ -60,24 +76,22 @@ def _call_api(self, video_id, title, key, data={}, **kwargs):
...
         subtitles = {}
-        for caption in set(re.findall(r'data-captions=\"([^\"]+)\"', webpage)):
-            renditions = self._call_api(
-                video_id, title, key, {'Captions': caption}, fatal=False,
-                note=f'Downloading {caption} subtitle information').get('renditions') or {}
-            for rendition_id, rendition in renditions.items():
-                self._extract_subtitles_from_rendition(rendition, subtitles, parsed_urls)
+        webpage = self._download_webpage(url, video_id, fatal=False) or ''
+
+        settings = self._call_api(video_id, title, key)
```
Comment (pukkandan) [inline yt_dlp/extractor/hidive.py:83]: Duplicate API call - Pass it into the function from _real_extract

File: yt_dlp/extractor/hidive.py [hunk 4]
```diff
@@ -46,8 +46,24 @@ def _perform_login(self, username, password):
...
+        #   support declaring a profile in a config file.
+        profileId = profileIds[0]
+        profileIdHash = profileIdHashes[0]
+        chooseProfileResultWebpage = self._download_json(
+            'https://www.hidive.com/ajax/chooseprofile', '',
+            data=urlencode_postdata({
+                'profileId': profileId,
+                'hash': profileIdHash,
+                'returnUrl': '/dashboard'
+            })) or {}
```
Comment (pukkandan) [inline yt_dlp/extractor/hidive.py:66]: snake case

File: yt_dlp/extractor/hidive.py [hunk 5]
```diff
@@ -46,8 +46,24 @@ def _perform_login(self, username, password):
...
             self._LOGIN_URL, None, 'Logging in', data=urlencode_postdata(data))
+        # If the user has multiple profiles, select one.
+        profileIds = re.findall(r'\<button .+data-profile-id="([a-zA-Z0-9]+)".*\>', loginWebpage)
+        if profileIds is None:
+            return # If only one profile, Hidive auto-selects it.
+        profileIdHashes = re.findall(r'\<button .+data-hash="([a-zA-Z0-9]+)".*\>', loginWebpage)
+        # Just pick the first profile, until someone else updates this code to
+        #   support declaring a profile in a config file.
+        profileId = profileIds[0]
+        profileIdHash = profileIdHashes[0]
```
Comment (pukkandan) [inline yt_dlp/extractor/hidive.py:59]: * Use `self._search_regex`
* Simplify regex (`a-zA-Z0-9` => `\w` etc)
* Avoid `.*`

File: yt_dlp/extractor/hidive.py [hunk 6]
```diff
@@ -46,8 +46,24 @@ def _perform_login(self, username, password):
...
+        chooseProfileResultWebpage = self._download_json(
+            'https://www.hidive.com/ajax/chooseprofile', '',
```
Comment (pukkandan) [inline yt_dlp/extractor/hidive.py:61]:
```suggestion
            'https://www.hidive.com/ajax/chooseprofile', None,
```

File: yt_dlp/extractor/hidive.py [hunk 7]
```diff
@@ -46,8 +46,24 @@ def _perform_login(self, username, password):
...
+        # If the user has multiple profiles, select one.
+        profileIds = re.findall(r'\<button .+data-profile-id="([a-zA-Z0-9]+)".*\>', loginWebpage)
+        if profileIds is None:
+            return # If only one profile, Hidive auto-selects it.
+        profileIdHashes = re.findall(r'\<button .+data-hash="([a-zA-Z0-9]+)".*\>', loginWebpage)
+        # Just pick the first profile, until someone else updates this code to
+        #   support declaring a profile in a config file.
+        profileId = profileIds[0]
+        profileIdHash = profileIdHashes[0]
+        chooseProfileResultWebpage = self._download_json(
```
Comment (pukkandan) [inline yt_dlp/extractor/hidive.py:60]: unsed?

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/hidive.py (hunk 2 above)
Comment (chexxor) [reply yt_dlp/extractor/hidive.py]: Thanks for the encouragement to look into that!
I improved the comments in the new commit I made.

File: yt_dlp/extractor/hidive.py (hunk 3 above)
Comment (chexxor) [reply yt_dlp/extractor/hidive.py]: Cool! I didn't know that was possible! I made this improvement in the new commit.

File: yt_dlp/extractor/hidive.py (hunk 1 above)
Comment (chexxor) [reply yt_dlp/extractor/hidive.py]: Nice find! I've removed

File: yt_dlp/extractor/hidive.py (hunk 5 above)
Comment (chexxor) [reply yt_dlp/extractor/hidive.py]: Thanks! I didn't know there was a preference for that function! I've changed to use that one.

**Comment by chexxor** [comment]
@pukkandan Your suggestions were excellent -- thanks for the code review! I learned a lot about Python while investigating and making those improvements.

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/hidive.py (hunk 2 above)
Comment (pukkandan) [reply yt_dlp/extractor/hidive.py]: I was asking to explain it to me - coz I don't understand this code/comment. You shouldn't use long name as language code. subtitles have a separate `name` field

**Comment by chexxor** [comment]
Thanks for your suggested changes! I think you're much more experienced with regex than I am -- I'll review your changes, apply it and test it to ensure it still works as expected.

**Comment by chexxor** [comment]
After I made the changes you suggested, I saw that GitHub has a button to commit your changes directly to my branch -- I didn't do that, so I'm sorry if that bothered you. I'm also not sure whether you prefer I click the "Resolve" button on the items you raised, or whether you do that.

I did test these changes, and it seems to still work as expected!

File: yt_dlp/extractor/hidive.py [hunk 8]
```diff
@@ -39,15 +39,31 @@ def _perform_login(self, username, password):
...
             self._LOGIN_URL, None, 'Logging in', data=urlencode_postdata(data))
+        # If the user has multiple profiles on their account, select one.
+        # For now pick the first profile. In the future, when someone has a use-case, they can update
+        #    this code to support that. Maybe the user would need to declare the index number of the profile
+        #    in a config file, and this part reads that config and selects the profile in the Nth position.
+        #    In that case, `_search_regex` likely wouldn't work, as it selects the first match.
+        profile_id = self._search_regex(r'\<button .+?data-profile-id="(\w+)".*?\>', login_webpage, 'profile_id')
+        if profile_id is None:
+            return # If only one profile, Hidive auto-selects it, so no more actions to take.
+        profile_id_hash = self._search_regex(r'\<button .+?data-hash="(\w+)".*?\>', login_webpage, 'profile_id_hash')
```
[the remark this answers is not in the document]
Comment (chexxor) [reply yt_dlp/extractor/hidive.py]: This regex is an interesting alternative! Thanks for sharing it -- I've never thought about using a regex like that before.

**Comment by pukkandan** [comment]
> After I made the changes you suggested, I saw that GitHub has a button to commit your changes directly to my branch -- I didn't do that, so I'm sorry if that bothered you.

Either is fine as long as the changes get made 👍 


> I'm also not sure whether you prefer I click the "Resolve" button on the items you raised, or whether you do that.

I'll resolve them

**Comment by pukkandan** [comment]
Pls reply to https://github.com/yt-dlp/yt-dlp/pull/5828#discussion_r1063746787

Everything else looks good

File: yt_dlp/extractor/hidive.py (hunk 2 above)
Comment (chexxor) [reply yt_dlp/extractor/hidive.py]: With it as-is, here's what --list-subs shows:
```
Language     Formats
english-caps vtt
english-subs vtt
```

I have removed that comment in the latest version of this file -- it didn't make sense to me, either. Instead, I listed out example values in that data structure.

Do you still want me to make any changes here? This comment is on an outdated version.

**Pull request merged**